### ISSUE TITLE
npm run develop results in address already in use 127.0.0.1:xxxx

### ISSUE THREAD

**Issue by sol-idsnake** [gatsbyjs/gatsby#30308]
This happens with the starter as well as customized apps.

Checking via `netstat -a` on Windows the ports in question all seem to be open and available, and although I am using the standard dev port 8000, I get the following error:

```
⠋  gatsby-source-wordpress  pull updates since last build
Error: listen EADDRINUSE: address already in use 127.0.0.1:63955
    at Server.setupListenHandle [as _listen2] (net.js:1318:16)
    at listenInCluster (net.js:1366:12)
    at GetAddrInfoReqWrap.doListen (net.js:1503:7)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:69:8)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1345:8)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {  
  code: 'EADDRINUSE',
  errno: -4091,
  syscall: 'listen',
  address: '127.0.0.1',
  port: 63955
}
```

This happens with any other manually entered port as well. A fresh install may work for a bit but eventually fails as well. I have predominantly noticed this after changing Wordpress data and restarting the process, but can't confirm that it only happens after changing data.

I am using a docker setup, but the problem persists when using Local by Flywheel

```
version: "3.8"

services:
  db:
    image: mariadb:10.5.8
    restart: always
    volumes:
      - db-data:/var/lib/mysql
    environment:
      MYSQL_ROOT_PASSWORD: root
      MYSQL_DATABASE: wordpress
      MYSQL_USER: wordpress
      MYSQL_PASSWORD: wordpress
    ports:
      - 3306:3306

  wordpress:
    depends_on:
      - db
    image: wordpress:latest
    ports:
      - "3000:80"
    restart: always
    environment:
      WORDPRESS_DB_HOST: db:3306
      WORDPRESS_DB_USER: wordpress
      WORDPRESS_DB_PASSWORD: wordpress
      WORDPRESS_DB_NAME: wordpress
    volumes:
      - wordpress:/var/www/html

  adminer:
    image: adminer
    restart: always
    ports:
      - 8080:8080

volumes:
  db-data:
  wordpress:
```
Gatsby config:

```
  plugins: [
    {
      resolve: "gatsby-source-wordpress-experimental",
      options: {
        url: "http://localhost:3000/graphql",
      },
     // removed the other plugins for brevity
  ],
```

And versions (again edited for brevity):

```
"dependencies": {
    "gatsby": "^2.31.1",
    "gatsby-source-wordpress-experimental": "^7.0.3",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
  },
```

I don't know where the process gets this port from 127.0.0.1:63955, since I either don't specify it or stick to something shorter (e.g. -p 4000). I've also noticed that once the develop script is done and the dev url is shown, the process starts over as if it has trouble persisting the changes.

If I am forgetting anything please let me know and I will edit the post promptly. I have found similar posts but they always showed that the specified port was in use.

**Comment by stevepepple** [comment]
I also see this issue intermittently with Node v14.15.1. 
I don't see any notable error during commands in gatsby node with the --verbose flag.
Server starts up and then stops. 
The `-H 0.0.0.0` appears to fix it, but I won't know for sure as it happens erratically. 

Other time I see this error, which is probably a separate issue: 

```
warn [gatsby-source-wordpress] Found a duplicate ID in WordPress - this means you will have fewer nodes in Gatsby than in WordPress. This will need to be resolved in WP by
identifying and fixing the underlying bug with your WP plugins or custom code.

 ERROR 

TypeError: Cannot read property 'databaseId' of undefined
```

**Comment by AlexMorrisonDev716** [comment]
@stevepepple I'm also having this issue with my site during development. Any luck? In the meantime, I've downgraded gatsby-source-wordpress to 4.1.0-next.1 and everything seems fine.

gatsby 3.2.1
gatsby-source-wordpress 5.2.1

**Comment by stevepepple** [comment]
@AlexMorrisonDev716, the -H 0.0.0.0 and running on different ports, did help for me. I also found that the duplicate ID issue happened when I had hard cached Wordpress changes and the GraphQL schema changes. Hope that helps

**Comment by emilstig** [comment]
I'm experiencing the same issue as @sol-idsnake with a fresh gatsby v3 install using the latest version of `gatsby-source-wordpress` with no other boilerplate files.

I have tried with node versions `10`, `11`, `12`, `13`, `14`, `15`. Used the `gatsby clean` command and running develop with `-H 0.0.0.0` on a different port than the default. Testing with both a local and external Wordpress install.

I have also noticed that the gatsby-source-wordpress process seems to start over once the develop script is done and the dev url is shown.

The steps and errors happening after develop process is done:
```
You can now view wordpress-gatsby in the browser.
⠀
  Local:            http://localhost:8001/
  On Your Network:  http://192.168.1.128:8001/
⠀
View GraphiQL, an in-browser IDE, to explore your site's data and schema
⠀
  Local:            http://localhost:8001/___graphql
  On Your Network:  http://192.168.1.128:8001/___graphql
⠀
Note that the development build is not optimized.
To create a production build, use gatsby build
⠀
success Building development bundle - 9.872s
success  gatsby-source-wordpress  fetch root fields - 1.514s
success  gatsby-source-wordpress  fetch root fields - 1.231s
success  gatsby-source-wordpress  pull updates since last build - 4.527s
success Checking for changed pages - 0.001s
success source and transform nodes - 4.566s
success building schema - 0.607s
info Total nodes: 152, SitePage nodes: 4 (use --verbose for breakdown)
success createPages - 0.003s
success Checking for changed pages - 0.001s
success update schema - 0.124s
success onPreExtractQueries - 0.001s
success extract queries from components - 0.123s
success write out requires - 0.003s
success run page queries - 0.010s - 1/1 97.64/s
info  gatsby-source-wordpress  Watching for WordPress changes
/path-to-project/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: listen EADDRINUSE: address already in use 127.0.0.1:55509
    at Server.setupListenHandle [as _listen2] (node:net:1310:16)
    at listenInCluster (node:net:1358:12)
    at GetAddrInfoReqWrap.doListen (node:net:1496:7)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:71:8)
Emitted 'error' event on Server instance at:
    at emitErrorNT (node:net:1337:8)
    at processTicksAndRejections (node:internal/process/task_queues:81:21) {
  code: 'EADDRINUSE',
  errno: -48,
  syscall: 'listen',
  address: '127.0.0.1',
  port: 55509
}
```

Gatsby config:

```
plugins: [
  {
    resolve: "gatsby-source-wordpress",
    options: {
      url: "https://url.com/graphql",
    },
  },
  "gatsby-plugin-sharp",
  "gatsby-transformer-sharp",
],
```

Dependencies:
```

"dependencies": {
    "gatsby": "^3.2.1",
    "gatsby-plugin-sass": "^4.2.0",
    "gatsby-plugin-sharp": "^3.2.0",
    "gatsby-source-wordpress": "^5.2.1",
    "gatsby-transformer-sharp": "^3.2.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "sass": "^1.32.8"
  }

```

**Comment by ali1059** [comment]
Any update on this issue ?

**Comment by jacksellwood** [comment]
Hi there, it looks like several users are impacted by this bug, we're prioritizing it and hope to have a fix in the next month.
Completely unrelated, but in the meantime we're working on a significant rewrite of CMS Preview 🎉

**Comment by franciscocobas** [comment]
I hope this fix come soon, its terrible to deal with this issue every time we need to start the develop environment. Sometimes I try 5 or 6 times until finally the server starts.

**Comment by crapsh00t** [comment]
@franciscocobas You can actually get the plugin to run? We have development halted by this bug.

**Comment by franciscocobas** [comment]
> @franciscocobas You can actually get the plugin to run? We have development halted by this bug.

After several tries I'm able to start the server.

**Comment by ali1059** [comment]
@franciscocobas In my case after running for few seconds it gives the same error

**Comment by TylerBarnes** [comment]
@crapsh00t , @franciscocobas , @ali1059 @AlexMorrisonDev716 (and anyone else) can you copy/paste the output of `gatsby info` here? I'm investigating this now but having a hard time reproducing it for some reason. It'd be helpful to try to spot any overlap between your setups.

**Comment by TylerBarnes** [comment]
btw you can get around this problem by setting `process.env.WP_DISABLE_POLLING=1` as that part of the codebase is what's triggering this issue.

**Comment by esizer** [comment]
> btw you can get around this problem by setting `process.env.WP_DISABLE_POLLING=1` as that var disables the part of the codebase that's triggering this issue.

Confirming that setting `process.env.WP_DISABLE_POLLING=1` is a valid work around.

`gatsby config` output:

```shell

System:
    OS: macOS 11.4
    CPU: (4) x64 Intel(R) Core(TM) i5-7400 CPU @ 3.00GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.13 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 91.0.4472.77
    Safari: 14.1.1
  npmPackages:
    gatsby: ^3.6.0 => 3.6.0
    gatsby-plugin-force-trailing-slashes: ^1.0.4 => 1.0.5
    gatsby-plugin-gatsby-cloud: ^2.6.0 => 2.6.0
    gatsby-plugin-google-analytics: ^3.2.0 => 3.6.0
    gatsby-plugin-image: ^1.6.0 => 1.6.0
    gatsby-plugin-manifest: ^3.6.0 => 3.6.0
    gatsby-plugin-offline: ^4.6.0 => 4.6.0
    gatsby-plugin-react-helmet: ^4.6.0 => 4.6.0
    gatsby-plugin-root-import: ^2.0.6 => 2.0.6
    gatsby-plugin-sharp: ^3.6.0 => 3.6.0
    gatsby-plugin-sitemap: ^4.2.0 => 4.2.0
    gatsby-plugin-styled-components: ^4.6.0 => 4.6.0
    gatsby-plugin-typescript: ^3.6.0 => 3.6.0
    gatsby-source-filesystem: ^3.6.0 => 3.6.0
    gatsby-source-wordpress: ^5.5.0 => 5.6.0
    gatsby-transformer-sharp: ^3.6.0 => 3.6.0
  ```

**Comment by crapsh00t** [comment]
> btw you can get around this problem by setting `process.env.WP_DISABLE_POLLING=1` as that var disables the part of the codebase that's triggering this issue.

Awesome! 👏🏻

**Comment by ali1059** [comment]
@TylerBarnes 

System:
    OS: Linux 5.8 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.17.0 - /usr/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.13 - /usr/bin/npm
  Browsers:
    Chrome: 91.0.4472.77
    Firefox: 88.0.1
  npmPackages:
    gatsby: ^3.1.2 => 3.6.1
    gatsby-background-image: ^1.5.0 => 1.5.3
    gatsby-plugin-emotion: ^6.1.0 => 6.6.0
    gatsby-plugin-image: ^1.1.2 => 1.6.0
    gatsby-plugin-preact: ^5.5.0 => 5.6.0
    gatsby-plugin-react-helmet: ^4.1.0 => 4.6.0
    gatsby-plugin-s3: ^0.3.8 => 0.3.8
    gatsby-plugin-sharp: ^3.1.2 => 3.6.0
    gatsby-source-filesystem: ^3.1.0 => 3.6.0
    gatsby-source-graphql: ^3.1.0 => 3.6.0
    gatsby-source-wordpress: ^5.6.0 => 5.6.0
    gatsby-transformer-sharp: ^3.1.0 => 3.6.0
  npmGlobalPackages:
    gatsby-cli: 3.6.0

**Comment by franciscocobas** [comment]
> @crapsh00t , @franciscocobas , @ali1059 @AlexMorrisonDev716 (and anyone else) can you copy/paste the output of `gatsby info` here? I'm investigating this now but having a hard time reproducing it for some reason. It'd be helpful to try to spot any overlap between your setups.

```
System:
    OS: macOS 11.3
    CPU: (8) x64 Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.1 - /usr/local/bin/node
    npm: 6.14.12 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 90.0.4430.212
    Safari: 14.1
  npmPackages:
    gatsby: ^3.5.0 => 3.5.0
    gatsby-cli: ^3.2.0 => 3.5.0
    gatsby-image: ^3.1.0 => 3.5.0
    gatsby-plugin-gatsby-cloud: ^2.5.0 => 2.5.0
    gatsby-plugin-google-analytics: ^3.5.0 => 3.5.0
    gatsby-plugin-google-tagmanager: ^3.5.0 => 3.5.0
    gatsby-plugin-image: ^1.3.0 => 1.5.0
    gatsby-plugin-react-helmet: ^4.3.0 => 4.5.0
    gatsby-plugin-sass: ^4.1.0 => 4.5.0
    gatsby-plugin-sharp: ^3.3.0 => 3.5.0
    gatsby-plugin-styled-components: ^4.1.0 => 4.5.0
    gatsby-source-filesystem: ^3.3.0 => 3.5.0
    gatsby-source-wordpress: ^5.5.0 => 5.5.0
    gatsby-transformer-sharp: ^3.3.0 => 3.5.0
  npmGlobalPackages:
    gatsby-cli: 3.1.0
```

**Comment by TylerBarnes** [comment]
I think I may have narrowed down what's causing this. I believe it's due to the WP plugin trying to source nodes before the dev server is ready 🤔 or atleast starting to watch for changes before the dev server is ready. I was able to reproduce by adding code to make it trigger node sourcing immediately when it started watching for changes. Can anyone who's running into this issue try this canary release and let me know if it fixes the problem? `gatsby-source-wordpress@5.7.0-alpha-wordpress.0+fe81b6e978`

Thanks!

**Comment by kelleyco2** [comment]
> I think I may have narrowed down what's causing this. I believe it's due to the WP plugin trying to source nodes before the dev server is ready 🤔 or at/near the moment that a code change is saved. I was able to reproduce by adding code to make it trigger node sourcing immediately when it started watching for changes. Can anyone who's running into this issue try this canary release and let me know if it fixes the problem? `gatsby-source-wordpress@5.7.0-alpha-wordpress.0+fe81b6e978`
> 
> Thanks!

@TylerBarnes This worked for me!! Thank you so much. This was causing quite the headache!

**Comment by TylerBarnes** [comment]
Woo :smile: that's great to hear! This fix should go out in the next Gatsby monorepo release but in the mean time it'd be great if others could try the canary and let me know if it's working for your site as well. :+1:

**Comment by ali1059** [comment]
@TylerBarnes This worked for me too. Thank you so much .

**Comment by franciscocobas** [comment]
> I think I may have narrowed down what's causing this. I believe it's due to the WP plugin trying to source nodes before the dev server is ready 🤔 or at/near the moment that a code change is saved. I was able to reproduce by adding code to make it trigger node sourcing immediately when it started watching for changes. Can anyone who's running into this issue try this canary release and let me know if it fixes the problem? `gatsby-source-wordpress@5.7.0-alpha-wordpress.0+fe81b6e978`
> 
> Thanks!

Sorry if I'm sounds a little bit silly, but how do I test with that release?

**Comment by TylerBarnes** [comment]
@franciscocobas definitely not a silly question :) You can install it like `yarn add gatsby-source-wordpress@5.7.0-alpha-wordpress.0+fe81b6e978` or `npm install gatsby-source-wordpress@5.7.0-alpha-wordpress.0+fe81b6e978`

**Comment by TylerBarnes** [comment]
I published another canary version that's exactly the same as the one above but with the debugging console.log's removed `gatsby-source-wordpress@5.7.0-alpha-wordpress.8+e70deff92e`.

**Comment by LekoArts** [comment]
Fixed by https://github.com/gatsbyjs/gatsby/pull/31713
